rust-runtime/inlineable/src/constrained.rs

```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+pub(crate) trait Constrained {
```

Comment:
Good question. I remember I went down this rabbit hole, but I forgot about it, so I went down again. This time I documented my efforts. I tried to do so in https://github.com/awslabs/smithy-rs/pull/1342/commits/1de207c39039ae614cdb5f813ed913c29b824de2.

However, as the commit message notes:

> it's not possible, because if we do, then we need both
`RuntimeType`s to declare a dependency on
`ServerCargoDependency.SmithyHttpServer`, but we can't do that in
`RuntimeType.kt`, which is in `codegen-core`, since we can't make
`codegen-core` take a dependency on `codegen-server`.
>
> But if we put them in `ServerRuntimeType`, we need to move the `data
class MaybeConstrained` (which would move the `MaybeConstrained`
`RuntimeType`) out of `RustTypes`, since `RustTypes` is in in
`codegen-core`. But `RustTypes` is `sealed`, and we can't inherit from a
sealed class outside the module where the class is defined. We'd have to
make `RustTypes` `abstract`. But then `RustType.render` can't be
exhaustive, and I don't know how to continue.

I reverted the commit, but I'd be happy if someone smarter than me with Kotlin could shuffle the right bits around to enable moving this to `aws-smithy-http-server`.